irc_client/commands.py

```diff
@@ -1,141 +1,111 @@
-        def msg_nickserv(message: str) -> str | None:
-            return msg("NickServ " + message)
-
-        @self.add_command("/ms")
-        @self.add_command("/memoserv")
-        def msg_memoserv(message: str) -> str | None:
-            return msg("MemoServ " + message)
-
-        # TODO: /me, /kick, /ban etc... lots of commands to add
-
-    def handle_command(
-        self, current_channel_or_nick: str | None, message: str
-    ) -> str | None:
-        """This runs when the user types something.
-
-        This takes care of special commands like ``/join #somechannel``
-        as well as sending messages to the channel.
-
-        The current_channel_or_nick can be:
-            * a nickname that the user is talking with using PMs
-            * a channel name
-            * None if the user is not looking at a channel right now
-        """
-        if not message:
-            return None
-
-        # '//lol' escapes the /, but '//lol/' is a literal '/lol/'
-        if message.startswith("/") and not message.startswith("//"):
-            # we have a special command
-            try:
-                command, args = message.split(maxsplit=1)
-            except ValueError:
-                command = message.strip()
-                args = ""
-
-            if command in self._commands:
-                return self._commands[command](args)
-            return f"There's no '{command}' command :("
-
-        # not a special command
-        if current_channel_or_nick is None:
-            return (
-                "You can't send messages directly to the server. "
-                "Join a channel instead and send messages there."
-            )
 
-        if message.startswith("//"):
-            message = message.replace("//", "/", 1)
-        self.irc_core.send_privmsg(current_channel_or_nick, message)
+_CommandT = TypeVar("_CommandT", bound=Callable[..., None])
+_commands: dict[str, tuple[str, Callable[..., None]]] = {}
+
+
+def add_command(usage: str) -> Callable[[_CommandT], _CommandT]:
+    assert re.fullmatch(r"/[a-z]+( <[a-z_]+>)*( \[<[a-z_]+>\])*", usage)
+
+    def do_it(func: _CommandT) -> _CommandT:
+        _commands[usage.split()[0]] = (usage, func)
+        return func
+
+    return do_it
+
+
+def handle_command(view: View, core: IrcCore, entry_content: str) -> None:
+    if not entry_content:
         return None
+
+    # TODO: disallow slashes in command
+    if entry_content.startswith("/") and not entry_content.startswith("//"):
+        try:
+            command, args_string = entry_content.split(maxsplit=1)
+        except ValueError:
+            command = entry_content.strip()
+            args_string = ""
+
+        if command in _commands:
+            usage, func = _commands[command]
+            required = usage.count(" <")
+            optional = usage.count(" [<")
+            # Last arg can contain spaces
+            args = args_string.split(maxsplit=required + optional - 1)
```

Comment:
This has a corner case when there's a command that takes 0 args, and `maxsplit=-1` is same as not specifying `maxsplit`